### PR TITLE
🦺 server: filter notifications to known tokens only

### LINKED ISSUE #906: server: filter spam notifications
https://exactlyprotocol.slack.com/archives/C05SNC75RPD/p1773956765585249

### PR DESCRIPTION
**Pull request by aguxez**
closes #906 

## Summary by CodeRabbit

* **Bug Fixes**
  * "Funds received" notifications now filter to recognized tokens only, significantly reducing spam from unknown or unverified assets
  * Enhanced error handling and logging for notification delivery to better track and diagnose failures

* **Tests**
  * Added comprehensive test coverage for token filtering logic, including caching behavior, multi-chain scenarios, and various error handling edge cases

### CODE REVIEW

File: server/hooks/activity.ts [hunk 1]
```diff
@@ -138,25 +140,65 @@ export default new Hono().post(
...
+                  await redis
+                    .multi()
+                    .del(key)
+                    .sadd(key, ...tokenList)
+                    .expire(key, 120)
+                    .exec();
+                  knownToken = tokenList.includes(address);
+                }
+              }
+            }
```
[the remark this answers is not in the document]
Comment (aguxez) [reply server/hooks/activity.ts]: we're only deploying this on chains where LiFi's token list is guaranteed non-empty.

File: server/hooks/activity.ts [hunk 2]
```diff
@@ -138,25 +140,65 @@ export default new Hono().post(
...
-                    language,
-                    assetSymbol ? `${amount} ${assetSymbol}` : amount,
-                  ]),
-                )
-              : assetSymbol,
-          },
-        ),
-      }).catch((error: unknown) => captureException(error));
+      let knownToken = true;
+      if (chain.id !== anvil.id) {
```
[the remark this answers is not in the document]
Comment (aguxez) [reply server/hooks/activity.ts]: intentional — we don't want to notify on native ETH transfers on non-exa chains

File: server/hooks/activity.ts [hunk 3]
```diff
@@ -138,25 +140,27 @@ export default new Hono().post(
...
-                  Object.entries(f(value)).map(([language, amount]) => [
-                    language,
-                    assetSymbol ? `${amount} ${assetSymbol}` : amount,
-                  ]),
-                )
-              : assetSymbol,
-          },
-        ),
-      }).catch((error: unknown) => captureException(error));
+      if (marketsByAsset.has(underlying) || (await isKnownToken(chain.id, underlying))) {
```
[the remark this answers is not in the document]
Comment (aguxez) [reply server/hooks/activity.ts]: I think this does not apply to the current webhook setup. Alchemy address activity webhooks are network-scoped, and this webhook is created with `network: network()`, where `network()` is derived from the same generated `chain.id` that provides `wethAddress`. So an Optimism deployment receives Optimism activity, and `WETH` is the Optimism WETH address.

The concern would be valid if this same endpoint/signing setup handled activity webhooks from multiple Alchemy networks. That is not the current setup, so I am leaving this behavior unchanged.

[the remark this answers is not in the document]
Comment (aguxez) [reply server/hooks/activity.ts]: I think this does not apply to the current webhook setup. Alchemy address activity webhooks are network-scoped, and this webhook is created with `network: network()`, where `network()` is derived from the same generated `chain.id` that provides `wethAddress`. So an Optimism deployment receives Optimism activity, and `WETH` is the Optimism WETH address.

The concern would be valid if this same endpoint/signing setup handled activity webhooks from multiple Alchemy networks. That is not the current setup, so I am leaving this behavior unchanged.

File: server/hooks/activity.ts [hunk 4]
```diff
@@ -330,3 +334,36 @@ findWebhook(({ webhook_type, webhook_url }) => webhook_type === "ADDRESS_ACTIVIT
...
+    const [[, isMember], [, count]] = v.parse(
+      v.tuple([v.tuple([v.null(), v.number()]), v.tuple([v.null(), v.number()])]),
+      await redis.pipeline().sismember(key, address).scard(key).exec(),
+    );
+    if (isMember) return true;
+    if (count > 0) return false;
+    const response = await fetch(`https://li.quest/v1/tokens?chains=${chainId}`, {
+      signal: AbortSignal.timeout(5000),
+    });
+    if (!response.ok) return true;
```
[the remark this answers is not in the document]
Comment (aguxez) [reply server/hooks/activity.ts]: Fixed. Non-2xx LiFi token-list responses now throw into the existing `isKnownToken` catch path, so they are captured with `{ level: "error" }` while preserving fail-open behavior. Added a regression test for a 503 response to verify the notification still sends and the error is reported.
